### PR TITLE
feat: add badges, emotes, and filters for suspicious messages

### PR DESCRIPTION
**Pull request by iProdigy**
Follow-up to #5056

image

Now the only remaining todo (for a separate PR) is rendering monitored messages based on pubsub instead of irc

### CODE REVIEW

File: src/Application.cpp [hunk 1]
```diff
@@ -507,18 +507,24 @@ void Application::initPubSub()
...
+                auto channel =
                     this->twitch->getChannelOrEmptyByID(action.channelID);
 
-                if (chan->isEmpty())
+                if (channel->isEmpty())
+                {
+                    return;
+                }
+
+                auto *chan = dynamic_cast<TwitchChannel *>(channel.get());
```
Comment (pajlada) [inline src/Application.cpp:518]: nit: rename to `twitchChannel`

File: src/providers/twitch/TwitchMessageBuilder.cpp [hunk 2]
```diff
@@ -1114,10 +1114,9 @@ Outcome TwitchMessageBuilder::tryAppendEmote(const EmoteName &name)
 }
 
 std::optional<EmotePtr> TwitchMessageBuilder::getTwitchBadge(
-    const Badge &badge) const
+    const Badge &badge, const TwitchChannel *chan)
```
Comment (pajlada) [inline src/providers/twitch/TwitchMessageBuilder.cpp:1117]: nit: Rename the `TwitchChannel*` parameter to `twitchChannel` and make this a static function in the anonymous namespace (within `namespace { ... }`) at the top of the file

File: src/providers/twitch/TwitchMessageBuilder.cpp [hunk 3]
```diff
@@ -1190,10 +1189,23 @@ void TwitchMessageBuilder::appendTwitchBadges()
 
     auto badgeInfos = TwitchMessageBuilder::parseBadgeInfoTag(this->tags);
     auto badges = this->parseBadgeTag(this->tags);
+    TwitchMessageBuilder::appendTwitchBadges(this, badges, badgeInfos,
+                                             this->twitchChannel);
+}
+
+void TwitchMessageBuilder::appendTwitchBadges(
```
Comment (pajlada) [inline src/providers/twitch/TwitchMessageBuilder.cpp:1196]: It feels weird to have a static function accept a class like this of it's own base type. I think this function should instead just be defined in TwitchMessageBuilder but outside of the class
also nit: rename `TwitchChannel*` parameter to `twitchChannel`

File: src/providers/twitch/pubsubmessages/LowTrustUsers.cpp [hunk 4]
```diff
@@ -21,8 +21,14 @@ PubSubLowTrustUsersMessage::PubSubLowTrustUsersMessage(const QJsonObject &root)
...
-        this->text =
-            data.value("message_content").toObject().value("text").toString();
+        const auto content = data.value("message_content").toObject();
+        this->text = content.value("text").toString();
+        std::vector<Fragment> parts;
+        for (const auto &part : content.value("fragments").toArray())
+        {
+            parts.emplace_back(part.toObject());
+        }
+        this->fragments = parts;
```
Comment (pajlada) [inline src/providers/twitch/pubsubmessages/LowTrustUsers.cpp:31]: We don't need the intermediate `parts` vector here, just emplace_back straight into `this->fragments`

File: src/providers/twitch/pubsubmessages/LowTrustUsers.cpp [hunk 5]
```diff
@@ -35,10 +41,12 @@ PubSubLowTrustUsersMessage::PubSubLowTrustUsersMessage(const QJsonObject &root)
...
             QColor(sender.value("chat_color").toString());
 
-        std::vector<LowTrustUserChatBadge> badges;
+        std::vector<Badge> badges;
         for (const auto &badge : sender.value("badges").toArray())
         {
-            badges.emplace_back(badge.toObject());
+            const auto badgeObj = badge.toObject();
+            badges.emplace_back(Badge{badgeObj.value("id").toString(),
+                                      badgeObj.value("version").toString()});
```
Comment (pajlada) [inline src/providers/twitch/pubsubmessages/LowTrustUsers.cpp:49]: nit: Make variables of the badge ID & badge version

**Review by pajlada — COMMENTED** [review]
Small nits before a more proper review

File: src/controllers/filters/lang/Filter.cpp [hunk 6]
```diff
@@ -101,6 +101,8 @@ ContextMap buildContextMap(const MessagePtr &m, chatterino::Channel *channel)
         {"flags.whisper", m->flags.has(MessageFlag::Whisper)},
         {"flags.reply", m->flags.has(MessageFlag::ReplyMessage)},
         {"flags.automod", m->flags.has(MessageFlag::AutoMod)},
+        {"flags.restricted", m->flags.has(MessageFlag::RestrictedMessage)},
+        {"flags.monitored", m->flags.has(MessageFlag::MonitoredMessage)},
```
Comment (dnsge) [inline src/controllers/filters/lang/Filter.cpp:105]: Also need to update a couple more things:

> 1. Update validIdentifiersMap in Tokenizer.hpp
> 2. Add the identifier to the list below
> 3. Add the type of the identifier to MESSAGE_TYPING_CONTEXT in Filter.hpp
> 4. Add the value for the identifier to the ContextMap returned by this function [[this change]]

**Review by pajlada — APPROVED** [review]

**Pull request merged**